### PR TITLE
Discussions and comment section (threads)

### PR DESCRIPTION
**Pull request by aapeliv**

### CODE REVIEW

**Comment by aapeliv** [comment]
Contrary to what I said in the meeting, I've instead gone the opposite and made the thread id compulsory. This makes the API a lot simpler to use, and allows using a truly "drop-in" comment section. I decided to give this a shot because the API for threads is such that it requires a lot of reworking to get the different entities that might have a comment section in there (because the thread_id is shared between the actual thread, comments and replies). In the end, an extra int foreign key is not very expensive.

This is up for debate, what do others think?

**Comment by kalvdans** [comment]
what is the difference between a discussion and a thread? can we merge them into one entity in the API?

**Comment by aapeliv** [comment]
> what is the difference between a discussion and a thread?

Discussion is a separate page for discussing things in a community or group. They will eventually be votable, I imagine, and have a bunch of other fancy features (for example, at the moment discussions have a slug and owners, threads don't). A thread is a "comment section", that can be embedded on any page.

> can we merge them into one entity in the API?

I think it's good to keep them separate, they will deviate over time, I believe.

**Comment by darrenvong** [comment]
Is a thread a recursive data structure with how it's currently modelled? i.e. a reply to a thread is a thread itself? Sorry if these might have been answered in previous discussions, I should keep up more with backend discussions too...

**Comment by aapeliv** [comment]
> Is a thread a recursive data structure with how it's currently modelled? i.e. a reply to a thread is a thread itself?

Yes, the API is recursive, in that you get and post onto `thread_id`s. But in fact there are only three levels: the root, comments, and replies. Replies therefore can't have replies on them.

**Comment by darrenvong** [comment]
Right, so I guess we are abandoning the Reddit style replies where things can nest to crazy levels for simplicity? Probably a good idea as I personally find heavily nested threads quite hard to follow

**Comment by aapeliv** [comment]
> Right, so I guess we are abandoning the Reddit style replies where things can nest to crazy levels for simplicity? Probably a good idea as I personally find heavily nested threads quite hard to follow

Yeah, that was the conclusion, I guess. I think the conclusion was that too deep nesting is just confusing!

File: app/frontend/src/components/Comments/CommentBox.tsx
Comment (lucaslcode) [inline app/frontend/src/components/Comments/CommentBox.tsx]:
```suggestion
  const handleComment = async (threadId: number, content: string) => {
```

File: app/frontend/src/components/Comments/CommentBox.tsx
Comment (lucaslcode) [inline app/frontend/src/components/Comments/CommentBox.tsx]: please change this to `replies` as children usually refers to react nodes.

File: app/frontend/src/components/Comments/CommentBox.tsx
Comment (lucaslcode) [inline app/frontend/src/components/Comments/CommentBox.tsx]:
```suggestion
              onComment={(content) => handleComment(comment.threadId, content)}
```
Convention is the event is `onEvent` and the function is `handleEvent` (though are other places in the code which don't currently reflect this)

File: app/frontend/src/components/Comments/NewComment.tsx
Comment (lucaslcode) [inline app/frontend/src/components/Comments/NewComment.tsx]:
```suggestion
  const handleSubmit= async () => {
```

File: app/frontend/src/components/Comments/NewComment.tsx
Comment (lucaslcode) [inline app/frontend/src/components/Comments/NewComment.tsx]:
```suggestion
      <Button onClick={handleSubmit} type="submit">Comment</Button>
```
Not 100% sure about `type="submit"` but do it if it lets you (good for a11y)

**Review by lucaslcode — CHANGES_REQUESTED** [review]
Just some minor style things (I'm assuming you clicked around to test)

**Comment by aapeliv** [comment]
@lucaslcode: fixed those comments. And yep, was just testing manually...

File: app/frontend/src/components/Comments/NewComment.tsx
Comment (darrenvong) [inline app/frontend/src/components/Comments/NewComment.tsx:37]: We should wrap this in a form and do it with `react-hook-form` like in other places, since it optimises for state change like this that otherwise would make this re-render on every keystroke and cause performance issues.

File: app/frontend/src/components/Comments/CommentBox.tsx
Comment (darrenvong) [inline app/frontend/src/components/Comments/CommentBox.tsx:42]: Probably ok for now... but definitely should do this in React Query to avoid having to deal with all the refetching ourselves. Is this meant to be working with pagination as well already?

**Review by darrenvong — CHANGES_REQUESTED** [review]

File: app/frontend/src/components/Comments/CommentBox.tsx
Comment (lucaslcode) [reply app/frontend/src/components/Comments/CommentBox.tsx:42]: > Is this meant to be working with pagination as well already?

Seems not, that's easier with react-query anyway

File: app/frontend/src/components/Comments/NewComment.tsx
Comment (lucaslcode) [reply app/frontend/src/components/Comments/NewComment.tsx:37]: @aapeliv if you make a note on #543 that this needs to be converted to react query and use react-hook-form, you can leave it for us to do

**Comment by lucaslcode** [comment]
I'm happy to approve this pending backend review. I will leave a note in #543 that we need to use `react-query` and `react-hook-forms`.

**Comment by aapeliv** [comment]
Yeah, the frontend is crappy (and also ugly). I'm hoping someone who actually knows frontend could fix this up in another PR, but wanted to do an "MVP" so the API calls are clear, etc, and hopefully it speeds that up!

I'm happy to also fix it up myself if yous want (though I can't make it pretty)!

File: app/backend/src/couchers/servicers/communities.py
Comment (kalvdans) [inline app/backend/src/couchers/servicers/communities.py:211]: Maybe a nicer error message if page_token can't be converted to integer. Can be done later though.

File: app/backend/src/couchers/servicers/discussions.py
Comment (kalvdans) [inline app/backend/src/couchers/servicers/discussions.py]: Ugly formatting, can you precalculate on a separate line or put parentheses around?

File: app/backend/src/tests/test_communities.py
Comment (kalvdans) [inline app/backend/src/tests/test_communities.py]: Drop the leading _, this is a test file anyway not exporting any functions.

File: app/backend/src/tests/test_discussions.py
Comment (kalvdans) [inline app/backend/src/tests/test_discussions.py]: Use Python's chained comparisons and have a more generous time limit, we don't want random failures because of loaded servers. You could set `test_start_time = now()` before creating the discussion and then change this to
```
assert test_start_time <= to_aware_datetime(res.created) <= now()
```

File: app/pb/communities.proto
Comment (kalvdans) [inline app/pb/communities.proto:73]: Would be nice to pass the number of comments here too, but could be added later.

File: app/pb/discussions.proto
Comment (kalvdans) [inline app/pb/discussions.proto:30]: Would be nice to pass number of comments here too, can be done later.

File: app/pb/groups.proto
Comment (kalvdans) [inline app/pb/groups.proto:79]: Likewise, pass number of comments but not necessary now.

**Review by kalvdans — DISMISSED** [review]
Backend part approved, some minor comments. You hereby get my auto-approval, no need to ask for another review if you fix those.

File: app/backend/src/couchers/servicers/communities.py
Comment (aapeliv) [reply app/backend/src/couchers/servicers/communities.py:211]: I think it's fine like this, it'll be clear from the logs and the exceptions are sanitized anyway at the end. Adding the check everywhere would be annoying and bloat the code, IMO. But we can refactor the page token stuff eventually to have shared helper functions or something, in which case it of course is worth it.

It's (almost certainly) a problem in something we're doing (or someone is poking around on the API) if a non-int ends up here. For the user, it'll just show up as "unknown internal error" or similar.

File: app/backend/src/couchers/servicers/discussions.py
Comment (aapeliv) [reply app/backend/src/couchers/servicers/discussions.py]: Agree that it's ugly, I'll see if I can make black format it nicer with parentheses.

Comment (aapeliv) [reply app/backend/src/couchers/servicers/discussions.py]: Fixed with precomputation

File: app/backend/src/tests/test_discussions.py
Comment (aapeliv) [reply app/backend/src/tests/test_discussions.py]: Thanks, great idea. I've cleaned it up

**Comment by aapeliv** [comment]
I've fixed up a bunch of stuff. Great idea with the timing.

> You hereby get my auto-approval

Thanks :D

Regarding getting number of comments: I think the API requires some refactoring, I want to push that out until a bit later when we see a bit more what the access patterns are, etc. At the moment the frontend has to make too many API calls for each page, and we need to figure out a long-term solutions for that. I've some ideas, but I don't want to jump in and guess, I want to have a good think about it.

**Pull request merged**
